File: packages/sdk-node/src/telemetry.ts
/**
 * Telemetry module for tracking client-side flag evaluations
 * and reporting them to the Rollgate server in batches.
 */

import { EventEmitter } from "events";

export interface TelemetryConfig {
  /** API endpoint for telemetry reporting */
  endpoint: string;
  /** API key for authentication */
  apiKey: string;
  /** Flush interval in milliseconds (default: 60000 = 1 minute) */
  flushIntervalMs: number;
  /** Maximum evaluations to buffer before forcing a flush (default: 1000) */
  maxBufferSize: number;
  /** Enable/disable telemetry (default: true) */
  enabled: boolean;
}

export const DEFAULT_TELEMETRY_CONFIG: TelemetryConfig = {
  endpoint: "",
  apiKey: "",
  flushIntervalMs: 60000, // 1 minute
  maxBufferSize: 1000,
  enabled: true,
};

export interface EvaluationStats {
  total: number;
  true: number;
  false: number;
}

export interface TelemetryPayload {
  evaluations: Record<string, EvaluationStats>;
  period_ms: number;
}

/**
 * TelemetryCollector tracks flag evaluations and sends them to the server in batches.
 * This is essential for client-side evaluation mode where the server doesn't see
 * individual evaluations.
 */
export class TelemetryCollector extends EventEmitter {
  private config: TelemetryConfig;
  private evaluations: Map<string, EvaluationStats> = new Map();
  private flushTimer: ReturnType<typeof setInterval> | null = null;
  private lastFlushTime: number = Date.now();
  private totalBuffered: number = 0;
  private isFlushing: boolean = false;

  constructor(config: Partial<TelemetryConfig> = {}) {
    super();
    this.config = { ...DEFAULT_TELEMETRY_CONFIG, ...config };
  }

  /**
   * Start the telemetry collector with periodic flushing
   */
  start(): void {
    if (!this.config.enabled || !this.config.endpoint || !this.config.apiKey) {
      return;
    }

    if (this.flushTimer) {
      return; // Already started
    }

    this.lastFlushTime = Date.now();
    this.flushTimer = setInterval(() => {
      this.flush().catch((err) => {
        this.emit("error", err);
      });
    }, this.config.flushIntervalMs);

    // Don't prevent Node.js from exiting
    if (this.flushTimer.unref) {
      this.flushTimer.unref();
    }
  }

  /**
   * Stop the telemetry collector and flush remaining data
   */
  async stop(): Promise<void> {
    if (this.flushTimer) {
      clearInterval(this.flushTimer);
      this.flushTimer = null;
    }

    // Final flush
    await this.flush();
  }

  /**
   * Record a flag evaluation
   */
  recordEvaluation(flagKey: string, result: boolean): void {
    if (!this.config.enabled) {
      return;
    }

    let stats = this.evaluations.get(flagKey);
    if (!stats) {
      stats = { total: 0, true: 0, false: 0 };
      this.evaluations.set(flagKey, stats);
    }

    stats.total++;
    if (result) {
      stats.true++;
    } else {
      stats.false++;
    }

    this.totalBuffered++;

    // Check if we need to flush due to buffer size
    if (this.totalBuffered >= this.config.maxBufferSize) {
      this.flush().catch((err) => {
        this.emit("error", err);
      });
    }
  }

  /**
   * Flush buffered evaluations to the server
   */
  async flush(): Promise<void> {
    if (this.isFlushing || this.evaluations.size === 0) {
      return;
    }

    if (!this.config.endpoint || !this.config.apiKey) {
      return;
    }

    this.isFlushing = true;

    // Capture current data and reset buffer
    const evaluationsToSend: Record<string, EvaluationStats> = {};
    for (const [key, stats] of this.evaluations) {
      evaluationsToSend[key] = { ...stats };
    }

    const periodMs = Date.now() - this.lastFlushTime;
    this.evaluations.clear();
    this.totalBuffered = 0;
    this.lastFlushTime = Date.now();

    const payload: TelemetryPayload = {
      evaluations: evaluationsToSend,
      period_ms: periodMs,
    };

    try {
      const response = await fetch(this.config.endpoint, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${this.config.apiKey}`,
        },
        body: JSON.stringify(payload),
      });

      if (!response.ok) {
        throw new Error(
          `Telemetry request failed: ${response.status} ${response.statusText}`,
        );
      }

      const result = (await response.json()) as { received: number };
      this.emit("flush", {
        flagsReported: Object.keys(evaluationsToSend).length,
        received: result.received,
        periodMs,
      });
    } catch (error) {
      // Put data back in buffer on failure
      for (const [key, stats] of Object.entries(evaluationsToSend)) {
        const existing = this.evaluations.get(key);
        if (existing) {
          existing.total += stats.total;
          existing.true += stats.true;
          existing.false += stats.false;
        } else {
          this.evaluations.set(key, stats);
        }
        this.totalBuffered += stats.total;
      }
      throw error;
    } finally {
      this.isFlushing = false;
    }
  }

  /**
   * Get current buffer stats (for debugging/monitoring)
   */
  getBufferStats(): { flagCount: number; evaluationCount: number } {
    return {
      flagCount: this.evaluations.size,
      evaluationCount: this.totalBuffered,
    };
  }

  /**
   * Update configuration
   */
  updateConfig(config: Partial<TelemetryConfig>): void {
    const wasEnabled = this.config.enabled;
    this.config = { ...this.config, ...config };

    // Restart if configuration changed significantly
    if (
      this.flushTimer &&
      config.flushIntervalMs &&
      config.flushIntervalMs !== this.config.flushIntervalMs
    ) {
      clearInterval(this.flushTimer);
      this.flushTimer = null;
      this.start();
    }

    // Start if newly enabled
    if (!wasEnabled && this.config.enabled) {
      this.start();
    }

    // Stop if disabled
    if (wasEnabled && !this.config.enabled && this.flushTimer) {
      clearInterval(this.flushTimer);
      this.flushTimer = null;
    }
  }
}

File: packages/sdk-go/telemetry.go
package rollgate

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"sync"
	"time"
)

// TelemetryConfig holds configuration for telemetry collection.
type TelemetryConfig struct {
	// FlushIntervalMs is the interval between automatic flushes (default: 60000ms)
	FlushIntervalMs int

	// MaxBufferSize is the maximum evaluations to buffer before forcing a flush (default: 1000)
	MaxBufferSize int

	// Enabled controls whether telemetry collection is active (default: true)
	Enabled bool
}

// DefaultTelemetryConfig returns default telemetry settings.
func DefaultTelemetryConfig() TelemetryConfig {
	return TelemetryConfig{
		FlushIntervalMs: 60000,
		MaxBufferSize:   1000,
		Enabled:         true,
	}
}

// TelemetryEvalStats tracks evaluation statistics for a single flag.
type TelemetryEvalStats struct {
	Total int `json:"total"`
	True  int `json:"true"`
	False int `json:"false"`
}

type telemetryPayload struct {
	Evaluations map[string]TelemetryEvalStats `json:"evaluations"`
	PeriodMs    int64                         `json:"period_ms"`
}

// TelemetryCollector tracks flag evaluations and sends them to the server in batches.
type TelemetryCollector struct {
	mu            sync.Mutex
	config        TelemetryConfig
	endpoint      string
	apiKey        string
	httpClient    *http.Client
	evaluations   map[string]*TelemetryEvalStats
	totalBuffered int
	lastFlushTime time.Time
	isFlushing    bool
	stopCh        chan struct{}
	stopped       bool
}

// NewTelemetryCollector creates a new telemetry collector.
func NewTelemetryCollector(endpoint, apiKey string, config TelemetryConfig, httpClient *http.Client) *TelemetryCollector {
	return &TelemetryCollector{
		config:        config,
		endpoint:      endpoint,
		apiKey:        apiKey,
		httpClient:    httpClient,
		evaluations:   make(map[string]*TelemetryEvalStats),
		lastFlushTime: time.Now(),
		stopCh:        make(chan struct{}),
	}
}

// Start begins periodic flushing.
func (tc *TelemetryCollector) Start() {
	if !tc.config.Enabled || tc.endpoint == "" || tc.apiKey == "" {
		return
	}

	interval := time.Duration(tc.config.FlushIntervalMs) * time.Millisecond
	go func() {
		ticker := time.NewTicker(interval)
		defer ticker.Stop()

		for {
			select {
			case <-tc.stopCh:
				return
			case <-ticker.C:
				_ = tc.Flush()
			}
		}
	}()
}

// Stop stops the telemetry collector and performs a final flush.
func (tc *TelemetryCollector) Stop() {
	tc.mu.Lock()
	if tc.stopped {
		tc.mu.Unlock()
		return
	}
	tc.stopped = true
	tc.mu.Unlock()

	close(tc.stopCh)
	_ = tc.Flush()
}

// RecordEvaluation records a single flag evaluation.
func (tc *TelemetryCollector) RecordEvaluation(flagKey string, result bool) {
	if !tc.config.Enabled {
		return
	}

	tc.mu.Lock()
	stats, ok := tc.evaluations[flagKey]
	if !ok {
		stats = &TelemetryEvalStats{}
		tc.evaluations[flagKey] = stats
	}

	stats.Total++
	if result {
		stats.True++
	} else {
		stats.False++
	}
	tc.totalBuffered++
	shouldFlush := tc.totalBuffered >= tc.config.MaxBufferSize
	tc.mu.Unlock()

	if shouldFlush {
		_ = tc.Flush()
	}
}

// Flush sends buffered evaluations to the server.
func (tc *TelemetryCollector) Flush() error {
	tc.mu.Lock()
	if tc.isFlushing || len(tc.evaluations) == 0 {
		tc.mu.Unlock()
		return nil
	}
	if tc.endpoint == "" || tc.apiKey == "" {
		tc.mu.Unlock()
		return nil
	}

	tc.isFlushing = true

	// Capture current data and reset buffer
	evaluationsToSend := make(map[string]TelemetryEvalStats, len(tc.evaluations))
	for key, stats := range tc.evaluations {
		evaluationsToSend[key] = *stats
	}
	periodMs := time.Since(tc.lastFlushTime).Milliseconds()
	tc.evaluations = make(map[string]*TelemetryEvalStats)
	tc.totalBuffered = 0
	tc.lastFlushTime = time.Now()
	tc.mu.Unlock()

	payload := telemetryPayload{
		Evaluations: evaluationsToSend,
		PeriodMs:    periodMs,
	}

	body, err := json.Marshal(payload)
	if err != nil {
		tc.restoreBuffer(evaluationsToSend)
		tc.mu.Lock()
		tc.isFlushing = false
		tc.mu.Unlock()
		return fmt.Errorf("marshal telemetry: %w", err)
	}

	req, err := http.NewRequest(http.MethodPost, tc.endpoint, bytes.NewReader(body))
	if err != nil {
		tc.restoreBuffer(evaluationsToSend)
		tc.mu.Lock()
		tc.isFlushing = false
		tc.mu.Unlock()
		return fmt.Errorf("create request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+tc.apiKey)

	resp, err := tc.httpClient.Do(req)
	if err != nil {
		tc.restoreBuffer(evaluationsToSend)
		tc.mu.Lock()
		tc.isFlushing = false
		tc.mu.Unlock()
		return fmt.Errorf("send telemetry: %w", err)
	}
	defer resp.Body.Close()

	tc.mu.Lock()
	tc.isFlushing = false
	tc.mu.Unlock()

	if resp.StatusCode != http.StatusOK {
		tc.restoreBuffer(evaluationsToSend)
		return fmt.Errorf("telemetry request failed: %d", resp.StatusCode)
	}

	return nil
}

// GetBufferStats returns current buffer statistics.
func (tc *TelemetryCollector) GetBufferStats() (flagCount, evaluationCount int) {
	tc.mu.Lock()
	defer tc.mu.Unlock()
	return len(tc.evaluations), tc.totalBuffered
}

func (tc *TelemetryCollector) restoreBuffer(data map[string]TelemetryEvalStats) {
	tc.mu.Lock()
	defer tc.mu.Unlock()
	for key, stats := range data {
		existing, ok := tc.evaluations[key]
		if ok {
			existing.Total += stats.Total
			existing.True += stats.True
			existing.False += stats.False
		} else {
			s := stats
			tc.evaluations[key] = &s
		}
		tc.totalBuffered += stats.Total
	}
}
